radixdlt-engine/src/main/java/com/radixdlt/atommodel/system/construction/NextEpochConstructorV3.java

```diff
@@ -0,0 +1,248 @@
+			);
+			i.forEachRemaining(preparedStake ->
+				map
+					.computeIfAbsent(
+						preparedStake.getDelegateKey(),
+						k -> new TreeMap<>((o1, o2) -> Arrays.compare(o1.getBytes(), o2.getBytes()))
+					)
+					.merge(preparedStake.getOwner(), preparedStake.getAmount(), UInt256::add)
+			);
+			return map;
+		});
+		for (var e : allPreparedUnstake.entrySet()) {
+			var k = e.getKey();
+			if (!validatorsToUpdate.containsKey(k)) {
+				var currentStake = txBuilder.down(
+					ValidatorStakeData.class,
+					p -> p.getValidatorKey().equals(k),
+					Optional.of(SubstateWithArg.noArg(ValidatorStakeData.create(k))),
+					"Validator not found"
+				);
+				validatorsToUpdate.put(k, currentStake);
+			}
+
+			var unstakes = e.getValue();
+			var curValidatorStake = validatorsToUpdate.get(k);
+			for (var entry : unstakes.entrySet()) {
+				var addr = entry.getKey();
+				var amt = entry.getValue();
+				var nextStakeAndAmt = curValidatorStake.unstakeOwnership(addr, amt, prevEpoch.getEpoch());
+				curValidatorStake = nextStakeAndAmt.getFirst();
+				var exittingStake = nextStakeAndAmt.getSecond();
+				txBuilder.up(exittingStake);
+			}
+			validatorsToUpdate.put(k, curValidatorStake);
+		}
+
+		var allPreparedStake = txBuilder.shutdownAll(PreparedStake.class, i -> {
+			i.forEachRemaining(preparedStake ->
+				preparingStake
+					.computeIfAbsent(
+						preparedStake.getDelegateKey(),
+						k -> new TreeMap<>((o1, o2) -> Arrays.compare(o1.getBytes(), o2.getBytes()))
+					)
+					.merge(preparedStake.getOwner(), preparedStake.getAmount(), UInt256::add)
+			);
+			return preparingStake;
+		});
+		for (var e : allPreparedStake.entrySet()) {
+			var k = e.getKey();
+			var stakes = e.getValue();
+			if (!validatorsToUpdate.containsKey(k)) {
+				var currentStake = txBuilder.down(
+					ValidatorStakeData.class,
+					p -> p.getValidatorKey().equals(k),
+					Optional.of(SubstateWithArg.noArg(ValidatorStakeData.create(k))),
+					"Validator not found"
+				);
+				validatorsToUpdate.put(k, currentStake);
+			}
+			var curValidator = validatorsToUpdate.get(k);
+			for (var entry : stakes.entrySet()) {
+				var addr = entry.getKey();
+				var amt = entry.getValue();
+
+				try {
+					var nextValidatorAndOwnership = curValidator.stake(addr, amt);
+					curValidator = nextValidatorAndOwnership.getFirst();
+					var stakeOwnership = nextValidatorAndOwnership.getSecond();
+					txBuilder.up(stakeOwnership);
+				} catch (ProcedureException ex) {
+					throw new TxBuilderException(ex);
+				}
+			}
+			validatorsToUpdate.put(k, curValidator);
+		}
+
+		var preparingRakeUpdates = new TreeMap<ECPublicKey, PreparedValidatorUpdate>(
+			(o1, o2) -> Arrays.compare(o1.getBytes(), o2.getBytes())
+		);
+		var buf = ByteBuffer.allocate(1 + Long.BYTES);
+		buf.put(SubstateTypeId.PREPARED_VALIDATOR_UPDATE.id());
+		buf.putLong(prevEpoch.getEpoch() + 1);
+		var index = new ShutdownAllIndex(buf.array(), PreparedValidatorUpdate.class);
```

Comment:
I'm a bit confused by the index usage. Are we saying `index = substate_type_id + <other_info>`?

Reply:
yup, exactly, it will actually be the first X bytes that represents the substate